File: feature/inCall/src/main/java/dev/alenajam/opendialer/feature/inCall/service/OngoingCall.java
package dev.alenajam.opendialer.feature.inCall.service;

import android.content.Context;
import android.net.Uri;
import android.os.Handler;
import android.os.Looper;
import android.provider.ContactsContract;
import android.telecom.Call;
import android.telecom.VideoProfile;

import java.util.List;

import javax.annotation.Nullable;

import dev.alenajam.opendialer.core.common.CommonUtils;
import dev.alenajam.opendialer.core.common.Contact;
import dev.alenajam.opendialer.feature.inCall.R;

public class OngoingCall {
    private static final int DTMF_DURATION_MS = 300;
    private final Call call;
    private String callerNumber = "", callerNumberLabel = "", callerName, callerImageUri = null;
    private long startTime = -1, totalTime = 0;
    private final Context context;
    private final CallsHandler callsHandler;
    private final long sequence;
    private final Handler mainHandler = new Handler(Looper.getMainLooper());
    private final Runnable stopDtmfTone;

    private final Call.Callback callback = new Call.Callback() {
        @Override
        public void onStateChanged(Call call, int newState) {
            super.onStateChanged(call, newState);
            updateState(newState);
            if (newState != Call.STATE_DISCONNECTED) callsHandler.updateCalls();
        }

        @Override
        public void onConferenceableCallsChanged(Call call, List<Call> conferenceableCalls) {
            super.onConferenceableCallsChanged(call, conferenceableCalls);
            callsHandler.updateCalls();
        }

        @Override
        public void onParentChanged(Call call, Call parent) {
            super.onParentChanged(call, parent);
            callsHandler.updateCalls();
        }

        @Override
        public void onDetailsChanged(Call call, Call.Details details) {
            super.onDetailsChanged(call, details);
            callsHandler.onCallDetailsChanged(OngoingCall.this);
        }
    };

    public OngoingCall(Context context, Call call, CallsHandler callsHandler, long sequence) {
        this.call = call;
        this.context = context;
        this.callsHandler = callsHandler;
        this.sequence = sequence;
        this.stopDtmfTone = call::stopDtmfTone;

        call.registerCallback(callback);
        refreshIdentity();
        updateState(getState());
    }

    public void refreshIdentity() {
        callerName = null;
        callerNumber = "";
        callerNumberLabel = "";
        callerImageUri = null;
        if (isConference()) {
            callerName = context.getString(R.string.conference_call);
        } else if (!isAnonymous()) {
            Uri numberUri = call.getDetails().getHandle();
            callerNumber = numberUri.getSchemeSpecificPart();
            callerName = callerNumber;
        }
    }

    public void applyContact(@Nullable Contact contact) {
        if (contact == null) return;
        callerName = contact.getName();
        callerImageUri = contact.getImageUri();
        callerNumberLabel = ContactsContract.CommonDataKinds.Phone.getTypeLabel(
                context.getResources(),
                contact.getPhoneType(),
                contact.getPhoneLabel()
        ).toString();
    }

    public void tearDown() {
        mainHandler.removeCallbacks(stopDtmfTone);
        call.stopDtmfTone();
        call.unregisterCallback(callback);
    }

    public void updateState(int state) {
        handleCall(state);
    }

    private void handleCall(int state) {
        switch (state) {
            case Call.STATE_HOLDING:
                long totalTime = CommonUtils.getCurrentTime() - getStartTime();
                totalTime += getTotalTime();
                this.totalTime = totalTime;
                break;
            case Call.STATE_DISCONNECTED:
                callsHandler.removeCall(call);
                OngoingCallHelper.handleDisconnectCause(context, call);
                break;
            case Call.STATE_ACTIVE:
                this.startTime = CommonUtils.getCurrentTime();
                break;
        }
    }

    public Call getCall() {
        return call;
    }

    public String getCallerNumber() {
        return callerNumber;
    }

    public String getCallerNumberLabel() {
        return callerNumberLabel;
    }

    @Nullable
    public String getCallerName() {
        return callerName;
    }

    @Nullable
    public String getCallerImageUri() {
        return callerImageUri;
    }

    public long getStartTime() {
        return startTime;
    }

    public long getTotalTime() {
        return totalTime;
    }

    public long getSequence() {
        return sequence;
    }

    public Integer getState() {
        if (call == null) return null;
        return call.getState();
    }

    public void answer() {
        if (call == null || getState() != Call.STATE_RINGING) return;
        call.answer(VideoProfile.STATE_AUDIO_ONLY);
    }

    public void hangup() {
        if (call == null) return;
        if (getState() == Call.STATE_RINGING) {
            call.reject(false, null);
        } else {
            call.disconnect();
        }
    }

    public void hangup(String message) {
        if (call == null) return;
        if (getState() == Call.STATE_RINGING) {
            call.reject(true, message);
        } else {
            call.disconnect();
        }
    }

    public void hold() {
        if (call == null) return;
        if (getState() == Call.STATE_HOLDING) {
            call.unhold();
        } else if (canBeHeld()) {
            call.hold();
        }
    }

    public void hold(boolean hold) {
        if (call == null) return;
        if (hold && canBeHeld()) call.hold();
        else if (!hold && getState() == Call.STATE_HOLDING) call.unhold();
    }

    public void playDtmf(char digit) {
        if (call == null || getState() != Call.STATE_ACTIVE) return;
        mainHandler.removeCallbacks(stopDtmfTone);
        call.stopDtmfTone();
        call.playDtmfTone(digit);
        mainHandler.postDelayed(stopDtmfTone, DTMF_DURATION_MS);
    }

    public boolean isAnonymous() {
        if (call == null) return false;
        return call.getDetails().getHandle() == null;
    }

    public boolean canBeMerged() {
        if (call == null) return false;

        if (call.getDetails().can(Call.Details.CAPABILITY_MERGE_CONFERENCE)) return true;

        if (call.getConferenceableCalls().size() > 0) return true;

        return false;
    }

    public boolean canBeHeld() {
        if (call == null) return false;
        return getState() == Call.STATE_HOLDING
                || call.getDetails().can(Call.Details.CAPABILITY_HOLD);
    }

    public boolean canBeSplit() {
        if (call == null) return false;
        return call.getDetails().can(Call.Details.CAPABILITY_SEPARATE_FROM_CONFERENCE);
    }

    public void split() {
        if (call == null || !canBeSplit()) return;
        call.splitFromConference();
    }

    public void merge() {
        if (call == null || !canBeMerged()) return;

        List<Call> conferenceableCalls = call.getConferenceableCalls();
        if (!conferenceableCalls.isEmpty()) {
            call.conference(conferenceableCalls.get(0));
        } else if (call.getDetails().can(Call.Details.CAPABILITY_MERGE_CONFERENCE)) {
            call.mergeConference();
        }
    }

    public boolean isConference() {
        if (call == null) return false;

        return call.getDetails().hasProperty(Call.Details.PROPERTY_CONFERENCE);
    }

    public boolean isConferenced() {
        return call != null && call.getParent() != null;
    }
}

File: feature/inCall/src/main/java/dev/alenajam/opendialer/feature/inCall/service/OngoingCall.kt
package dev.alenajam.opendialer.feature.inCall.service

import android.content.Context
import android.provider.ContactsContract
import android.telecom.Call
import android.telecom.VideoProfile
import dev.alenajam.opendialer.core.common.CommonUtils
import dev.alenajam.opendialer.core.common.Contact
import dev.alenajam.opendialer.feature.inCall.R
import kotlin.time.Duration.Companion.milliseconds
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.cancel
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch

data class OngoingCallState(
    val callerNumber: String = "",
    val callerNumberLabel: String = "",
    val callerName: String? = null,
    val callerImageUri: String? = null,
    val startTime: Long = -1,
    val totalTime: Long = 0,
    val state: Int = Call.STATE_NEW,
    val isConference: Boolean = false,
    val isConferenced: Boolean = false,
    val canMerge: Boolean = false,
    val canHold: Boolean = false,
    val canSplit: Boolean = false,
    val isAnonymous: Boolean = false
)

class OngoingCall(
    private val context: Context,
    val call: Call,
    private val onRemoved: (Call) -> Unit,
    val sequence: Long
) {
    private val scope = CoroutineScope(SupervisorJob() + Dispatchers.Main)
    private val _state = MutableStateFlow(OngoingCallState())
    val stateFlow: StateFlow<OngoingCallState> = _state.asStateFlow()

    companion object {
        private const val DTMF_DURATION_MS = 300L
    }

    private var lastState = Call.STATE_NEW
    private val missedCallTracker = MissedCallTracker()
    private var dtmfJob: Job? = null

    private val callback = object : Call.Callback() {
        override fun onStateChanged(call: Call, newState: Int) {
            super.onStateChanged(call, newState)
            updateCallState(newState)
        }

        override fun onConferenceableCallsChanged(call: Call, conferenceableCalls: List<Call>) {
            super.onConferenceableCallsChanged(call, conferenceableCalls)
            syncState()
        }

        override fun onParentChanged(call: Call, parent: Call?) {
            super.onParentChanged(call, parent)
            syncState()
        }

        override fun onDetailsChanged(call: Call, details: Call.Details) {
            super.onDetailsChanged(call, details)
            syncState()
        }
    }

    init {
        call.registerCallback(callback)
        refreshIdentity()
        updateCallState(call.state)
    }

    fun refreshIdentity() {
        val isConf = call.details.hasProperty(Call.Details.PROPERTY_CONFERENCE)
        val handle = call.details.handle
        val isAnon = handle == null
        val isVoicemail = handle?.scheme.equals("voicemail", ignoreCase = true)

        _state.update {
            it.copy(
                isConference = isConf,
                isAnonymous = isAnon,
                callerName = when {
                    isConf -> context.getString(R.string.conference_call)
                    isVoicemail -> context.getString(R.string.voicemail)
                    !isAnon -> handle?.schemeSpecificPart
                    else -> null
                },
                callerNumber = if (!isConf && !isAnon && !isVoicemail) handle?.schemeSpecificPart ?: "" else "",
                callerNumberLabel = "",
                callerImageUri = null
            )
        }
        syncState()
    }

    fun applyContact(contact: Contact?) {
        if (contact == null) return
        _state.update {
            it.copy(
                callerName = contact.name,
                callerImageUri = contact.imageUri,
                callerNumberLabel = ContactsContract.CommonDataKinds.Phone.getTypeLabel(
                    context.resources,
                    contact.phoneType,
                    contact.phoneLabel
                ).toString()
            )
        }
    }

    private fun syncState() {
        _state.update {
            it.copy(
                isConferenced = call.parent != null,
                canMerge = call.details.can(Call.Details.CAPABILITY_MERGE_CONFERENCE) || call.conferenceableCalls.isNotEmpty(),
                canHold = call.state == Call.STATE_HOLDING || call.details.can(Call.Details.CAPABILITY_HOLD),
                canSplit = call.details.can(Call.Details.CAPABILITY_SEPARATE_FROM_CONFERENCE),
                state = call.state
            )
        }
    }

    fun tearDown() {
        dtmfJob?.cancel()
        call.stopDtmfTone()
        call.unregisterCallback(callback)
        scope.cancel()
    }

    private fun updateCallState(state: Int) {
        if (state == lastState) return
        missedCallTracker.onStateChanged(state)

        when (state) {
            Call.STATE_HOLDING -> {
                if (lastState == Call.STATE_ACTIVE) {
                    accumulateActiveTime()
                }
            }
            Call.STATE_DISCONNECTED -> {
                if (lastState == Call.STATE_ACTIVE) {
                    accumulateActiveTime()
                }
                lastState = state
                onRemoved(call)
                OngoingCallHelper.handleDisconnectCause(context, call)
                return
            }
            Call.STATE_ACTIVE -> {
                _state.update { it.copy(startTime = CommonUtils.getCurrentTime()) }
            }
        }
        lastState = state
        syncState()
    }

    private fun accumulateActiveTime() {
        val current = _state.value
        if (current.startTime < 0) return
        val newTotal = current.totalTime + (CommonUtils.getCurrentTime() - current.startTime)
        _state.update { it.copy(totalTime = newTotal, startTime = -1) }
    }

    // Direct access to state for legacy/convenience
    val state: Int get() = _state.value.state
    val callerNumber: String get() = _state.value.callerNumber
    val callerName: String? get() = _state.value.callerName
    val callerNumberLabel: String get() = _state.value.callerNumberLabel
    val callerImageUri: String? get() = _state.value.callerImageUri
    val isConference: Boolean get() = _state.value.isConference
    val isConferenced: Boolean get() = _state.value.isConferenced
    val startTime: Long get() = _state.value.startTime
    val totalTime: Long get() = _state.value.totalTime
    val shouldNotifyMissedCall: Boolean get() = missedCallTracker.shouldNotify()

    fun answer() {
        if (!CallActionPolicy.shouldAnswer(state)) return
        call.answer(VideoProfile.STATE_AUDIO_ONLY)
    }

    fun hangup(message: String? = null) {
        if (CallActionPolicy.shouldReject(state)) {
            missedCallTracker.markLocallyDeclined()
            call.reject(message != null, message)
        } else {
            call.disconnect()
        }
    }

    fun markLocallyDeclined() {
        missedCallTracker.markLocallyDeclined()
    }

    fun hold(hold: Boolean? = null) {
        when (CallActionPolicy.holdAction(state, canBeHeld(), hold)) {
            CallActionPolicy.HoldAction.HOLD -> call.hold()
            CallActionPolicy.HoldAction.UNHOLD -> call.unhold()
            CallActionPolicy.HoldAction.NONE -> Unit
        }
    }

    fun canBeHeld(): Boolean = _state.value.canHold
    fun canBeMerged(): Boolean = _state.value.canMerge

    fun playDtmf(digit: Char) {
        if (!CallActionPolicy.canPlayDtmf(state)) return
        dtmfJob?.cancel()
        call.stopDtmfTone()
        call.playDtmfTone(digit)
        dtmfJob = scope.launch {
            delay(DTMF_DURATION_MS.milliseconds)
            call.stopDtmfTone()
        }
    }

    fun split() {
        if (!_state.value.canSplit) return
        call.splitFromConference()
    }

    fun merge() {
        if (!_state.value.canMerge) return
        val conferenceableCalls = call.conferenceableCalls
        if (conferenceableCalls.isNotEmpty()) {
            call.conference(conferenceableCalls[0])
        } else if (call.details.can(Call.Details.CAPABILITY_MERGE_CONFERENCE)) {
            call.mergeConference()
        }
    }
}
